model/link_metadata.go

```diff
@@ -0,0 +1,150 @@
+	// - nil if the linked content has no metadata
+	Data interface{}
+}
+
+func (o *LinkMetadata) PreSave() {
+	o.Hash = GenerateLinkMetadataHash(o.URL, o.Timestamp)
+}
+
+func (o *LinkMetadata) IsValid() *AppError {
+	if o.URL == "" {
+		return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.url.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	if o.Timestamp == 0 || !isRoundedToNearestHour(o.Timestamp) {
+		return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.timestamp.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	switch o.Type {
+	case LINK_METADATA_TYPE_IMAGE:
+		if o.Data == nil {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		if _, ok := o.Data.(*PostImage); !ok {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data_type.app_error", nil, "", http.StatusBadRequest)
+		}
+	case LINK_METADATA_TYPE_NONE:
+		if o.Data != nil {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data_type.app_error", nil, "", http.StatusBadRequest)
+		}
+	case LINK_METADATA_TYPE_OPENGRAPH:
+		if o.Data == nil {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		if _, ok := o.Data.(*opengraph.OpenGraph); !ok {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data_type.app_error", nil, "", http.StatusBadRequest)
+		}
+	default:
+		return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.type.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	return nil
+}
+
+// FixData converts o.Data from JSON into properly structured data. This is intended to be used after deserializing
+// a LinkMetadata object that has been stored in the database.
+func (o *LinkMetadata) FixData() error {
+	var b []byte
+	switch t := o.Data.(type) {
+	case []byte:
+		// MySQL uses a byte slice for JSON
+		b = t
+	case string:
+		// Postgres uses a string for JSON
+		b = []byte(t)
+	}
+
+	if b == nil {
+		// Data doesn't need to be fixed
+		return nil
+	}
+
+	var data interface{}
+	var err error
+
+	switch o.Type {
+	case LINK_METADATA_TYPE_IMAGE:
+		image := &PostImage{}
+
+		err = json.Unmarshal(b, &image)
+
+		data = image
+	case LINK_METADATA_TYPE_OPENGRAPH:
+		og := &opengraph.OpenGraph{}
+
+		json.Unmarshal(b, &og)
+
+		data = og
+	}
+
+	if err != nil {
+		return err
+	}
+
+	o.Data = data
+
+	return nil
+}
+
+// FloorToNearestHour takes a timestamp (in milliseconds) and returns it rounded to the previous hour.
+func FloorToNearestHour(ms int64) int64 {
+	t := time.Unix(0, ms*int64(1000*1000))
+
+	return time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), 0, 0, 0, t.Location()).UnixNano() / int64(time.Millisecond)
+}
+
+// isRoundedToNearestHour returns true if the given timestamp (in milliseconds) has been rounded to the nearest hour.
+func isRoundedToNearestHour(ms int64) bool {
```

Comment:
Maybe for simplicity this can be `return FloorToNearestHour(ms) == ms`

Reply:
Makes sense